patches/server/0981-Brigadier-based-command-API.patch

```diff
@@ -0,0 +1,2166 @@
++ */
++public class BukkitBrigForwardingMap extends HashMap<String, Command> {
++
++    private final EntrySet entrySet = new EntrySet();
++    private final KeySet keySet = new KeySet();
++    private final Values values = new Values();
++
++    // Previous dispatcher used to get commands to migrate to another dispatcher
++
++    CommandDispatcher<CommandSourceStack> getDispatcher() {
++        return Bukkit.getServer().getCommandDispatcher();
++    }
++
++    @Override
++    public int size() {
++        return this.getDispatcher().getRoot().getChildren().size();
++    }
++
++    @Override
++    public boolean isEmpty() {
++        return this.size() != 0;
++    }
++
++    @Override
++    public boolean containsKey(Object key) {
++        if (!(key instanceof String stringKey)) {
++            return false;
++        }
++
++        // Do any children match?
++        return this.getDispatcher().getRoot().getChild(stringKey) != null;
++    }
++
++    @Override
++    public boolean containsValue(Object value) {
++        if (value == null) {
++            return false;
++        }
++
++        for (CommandNode<CommandSourceStack> child : this.getDispatcher().getRoot().getChildren()) {
++            // If child is a bukkit command node, we can convert it!
++            if (child instanceof BukkitCommandNode bukkitCommandNode) {
++                return bukkitCommandNode.getBukkitCommand().equals(value);
++            }
++        }
++
++        return false;
++    }
++
++    @Override
++    public Command get(Object key) {
++        for (CommandNode<CommandSourceStack> child : this.getDispatcher().getRoot().getChildren()) {
++            if (child instanceof LiteralCommandNode<CommandSourceStack> literalCommandNode && literalCommandNode.getLiteral().equals(key)) {
++                if (literalCommandNode instanceof BukkitCommandNode bukkitCommandNode) {
++                    return bukkitCommandNode.getBukkitCommand();
++                }
++
++                return PaperBrigadier.wrapNode(child);
++            }
++        }
++
++
++        return null;
++    }
++
++    @Nullable
++    @Override
++    public Command put(String key, Command value) {
++        Command old = this.get(key);
++        this.getDispatcher().getRoot().removeCommand(key); // Override previous command
++        this.getDispatcher().getRoot().addChild(BukkitCommandNode.of(key, value));
++        return old;
++    }
++
++    @Override
++    public Command remove(Object key) {
++        if (!(key instanceof String string)) {
++            return null;
++        }
++
++        Command old = this.get(key);
++         if (old != null) {
++             this.getDispatcher().getRoot().removeCommand(string);
++         }
++
++        return old;
++    }
++
++    @Override
++    public boolean remove(Object key, Object value) {
++        Command old = this.remove(key);
++        return Objects.equals(old, value);
```

Comment:
This is incorrect for the Map#remove(Object, Object) contract. You only remove it if the value is equal.